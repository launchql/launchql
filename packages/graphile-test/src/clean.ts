const uuidRegexp = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

const idReplacement = (v: unknown): string | unknown => (!v ? v : '[ID]');

// Generic object type for any key-value mapping
type AnyObject = Record<string, any>;

function mapValues<T extends AnyObject, R = any>(
  obj: T,
  fn: (value: T[keyof T], key: keyof T) => R
): Record<keyof T, R> {
  return Object.entries(obj).reduce((acc, [key, value]) => {
    acc[key as keyof T] = fn(value, key as keyof T);
    return acc;
  }, {} as Record<keyof T, R>);
}

export const pruneDates = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) => {
    if (!v) {
      return v;
    }
    if (v instanceof Date) {
      return '[DATE]';
    } else if (
      typeof v === 'string' &&
      /(_at|At)$/.test(k as string) &&
      /^20[0-9]{2}-[0-9]{2}-[0-9]{2}/.test(v)
    ) {
      return '[DATE]';
    }
    return v;
  });

export const pruneIds = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) =>
    (k === 'id' || (typeof k === 'string' && k.endsWith('_id'))) &&
    (typeof v === 'string' || typeof v === 'number')
      ? idReplacement(v)
      : v
  );

export const pruneIdArrays = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) =>
    typeof k === 'string' && k.endsWith('_ids') && Array.isArray(v)
      ? `[UUIDs-${v.length}]`
      : v
  );

export const pruneUUIDs = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) => {
    if (typeof v !== 'string') {
      return v;
    }
    if (['uuid', 'queue_name'].includes(k as string) && uuidRegexp.test(v)) {
      return '[UUID]';
    }
    if (k === 'gravatar' && /^[0-9a-f]{32}$/i.test(v)) {
      return '[gUUID]';
    }
    return v;
  });

export const pruneHashes = (row: AnyObject): AnyObject =>
  mapValues(row, (v, k) =>
    typeof k === 'string' &&
    k.endsWith('_hash') &&
    typeof v === 'string' &&
    v.startsWith('$')
      ? '[hash]'
      : v
  );

export const prune = (obj: AnyObject): AnyObject =>
  pruneHashes(pruneUUIDs(pruneIds(pruneIdArrays(pruneDates(obj)))));

export const snapshot = (obj: unknown): unknown => {
  if (Array.isArray(obj)) {
    return obj.map(snapshot);
  } else if (obj && typeof obj === 'object') {
    return mapValues(prune(obj as AnyObject), snapshot);
  }
  return obj;
};
