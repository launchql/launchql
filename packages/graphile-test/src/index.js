export * from './graphql';
export * from './gql';
export * from './env';
export * from './clean';
