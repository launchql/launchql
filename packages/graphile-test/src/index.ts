export * from './clean';
export * from './graphile-test';
export * from './connect';
export * from './types';
export { seed } from 'pgsql-test';