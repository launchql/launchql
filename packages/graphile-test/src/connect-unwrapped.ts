import { getConnections as getPgConnections } from 'pgsql-test';
import { GraphQLTest } from './graphile-test';

import type { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';
import type { PgTestClient } from 'pgsql-test/test-client';
import type { SeedAdapter } from 'pgsql-test/seed/types';
import type {
  GraphQLQueryOptions,
  GraphQLQueryUnwrappedFn,
  GraphQLQueryUnwrappedFnPos,
  GraphQLResponse
} from './types';

export interface GetConnectionsInput {
  useRoot?: boolean;
  schemas: string[];
  authRole?: string;
}

const unwrap = <T>(res: GraphQLResponse<T>): T => {
  if (res.errors?.length) {
    throw new Error(JSON.stringify(res.errors, null, 2));
  }
  if (!res.data) {
    throw new Error('No data returned from GraphQL query');
  }
  return res.data;
};

export const getConnectionsUnwrapped = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryUnwrappedFn;
  queryPositional: GraphQLQueryUnwrappedFnPos;
}> => {
  const conn: GetConnectionResult = await getPgConnections(input, seedAdapters);
  const { pg, db, teardown: dbTeardown } = conn;

  const gqlContext = GraphQLTest(input, conn);
  await gqlContext.setup();

  const teardown = async () => {
    await gqlContext.teardown();
    await dbTeardown();
  };

  const query: GraphQLQueryUnwrappedFn = async (opts) => unwrap(await gqlContext.query(opts));

  const queryPositional: GraphQLQueryUnwrappedFnPos = async (query, variables, commit, reqOptions) =>
    unwrap(await gqlContext.query({ query, variables, commit, reqOptions }));

  return {
    pg,
    db,
    teardown,
    query,
    queryPositional
  };
};
