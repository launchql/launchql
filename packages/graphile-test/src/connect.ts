import { getConnections as getPgConnections } from 'pgsql-test';
import { GraphQLTest } from './graphile-test';
import type { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';
import type { PgTestClient } from 'pgsql-test/test-client';
import type { SeedAdapter } from 'pgsql-test/seed/types';
import type { DocumentNode, ExecutionResult } from 'graphql';

export type GraphQLQueryFn<T = ExecutionResult> = (
  query: string | DocumentNode,
  variables?: Record<string, any>,
  commit?: boolean,
  reqOptions?: Record<string, any>
) => Promise<T>;
export interface GetConnectionsInput {
  useRoot?: boolean;
  schemas: string[];
  authRole?: string;
}

/**
 * Combines PostgreSQL test setup with GraphQL test context
 */
export const getConnections = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFn;
}> => {
  
  const conn: GetConnectionResult = await getPgConnections(input, seedAdapters);
  const { pg, db, teardown: dbTeardown } = conn;

  const gqlContext = GraphQLTest(input, conn);
  
  await gqlContext.setup();

  const teardown = async () => {
    await gqlContext.teardown();
    await dbTeardown();
  };

  const query: GraphQLQueryFn = (query, variables, commit, reqOptions) =>
    gqlContext.query({ query, variables, commit, reqOptions });

  return {
    pg,
    db,
    teardown,
    query
  };
};
