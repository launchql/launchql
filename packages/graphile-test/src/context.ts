import { graphql, print, ExecutionResult, DocumentNode } from 'graphql';
import { withPostGraphileContext, PostGraphileOptions } from 'postgraphile';
// @ts-ignore
import MockReq from 'mock-req';
import type { Client, Pool } from 'pg';
import { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';
import { GetConnectionsInput } from './connect';

interface PgSettings {
    [key: string]: string;
}

export const runGraphQLInContext = async <T = ExecutionResult>({
    input,
    conn,
    pgPool,
    schema,
    options,
    authRole,
    query,
    variables,
    reqOptions = {}
}: {
    input: GetConnectionsInput & GetConnectionOpts,
    conn: GetConnectionResult;
    pgPool: Pool;
    schema: any;
    options: PostGraphileOptions;
    authRole: string;
    query: string | DocumentNode;
    variables?: Record<string, any>;
    reqOptions?: Record<string, any>;
}): Promise<T> => {
    if (!conn.pg.client) {
        throw new Error('pgClient is required and must be provided externally.');
    }

    const req = new MockReq({
        url: options.graphqlRoute || '/graphql',
        method: 'POST',
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json'
        },
        ...reqOptions
    });

    const pgSettingsGenerator = options.pgSettings;
    // @ts-ignore
    const pgSettings: PgSettings =
        typeof pgSettingsGenerator === 'function'
            ? await pgSettingsGenerator(req)
            : pgSettingsGenerator || {};

    // @ts-ignore
    return await withPostGraphileContext(   
        { ...options, pgPool, pgSettings },
        async context => {

            const pgConn = input.useRoot ? conn.pg : conn.db;
            const pgClient = pgConn.client;
            await setContextOnClient(pgClient, pgSettings, authRole);
            await pgConn.ctxQuery();

            const printed = typeof query === 'string' ? query : print(query);
            const result = await graphql({
                schema,
                source: printed,
                contextValue: { ...context, pgClient },
                variableValues: variables ?? null
            });
            return result as T;
        }
    );
};


export async function setContextOnClient(
  pgClient: Client,
  pgSettings: Record<string, string>,
  role: string
): Promise<void> {
  await pgClient.query(`select set_config('role', $1, true)`, [role]);

  for (const [key, value] of Object.entries(pgSettings)) {
    await pgClient.query(`select set_config($1, $2, true)`, [key, String(value)]);
  }
}
