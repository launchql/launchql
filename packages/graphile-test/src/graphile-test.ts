import type { GraphQLSchema } from 'graphql';
import type { GraphQLTestContext, GraphQLQueryOptions } from './types';

import { runGraphQLInContext } from './context';
import { createPostGraphileSchema, PostGraphileOptions } from 'postgraphile';
import { getGraphileSettings } from 'graphile-settings';
import { GetConnectionsInput } from './types';
import { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';

export const GraphQLTest = (
  input: GetConnectionsInput & GetConnectionOpts,
  conn: GetConnectionResult
): GraphQLTestContext => {
  const {
    schemas,
    authRole
  } = input;

  let schema: GraphQLSchema;
  let options: PostGraphileOptions;

  const pgPool = conn.manager.getPool(conn.pg.config);

  const setup = async () => {
    options = getGraphileSettings({ graphile: { schema: schemas } });
    schema = await createPostGraphileSchema(pgPool, schemas, options);
  };

  const teardown = async () => { /* optional cleanup */ };

  const query = async <TResult = any, TVariables = Record<string, any>>(
    opts: GraphQLQueryOptions<TVariables>
  ): Promise<TResult> => {
    return await runGraphQLInContext<TResult>({
      input,
      schema,
      options,
      authRole,
      pgPool,
      conn,
      ...opts
    });
  };

  return {
    setup,
    teardown,
    query
  };
};
