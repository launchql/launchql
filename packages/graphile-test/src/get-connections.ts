import { getConnections as getPgConnections } from 'pgsql-test';
import { GraphQLTest } from './graphile-test';

import type { GetConnectionOpts, GetConnectionResult } from 'pgsql-test';
import type { PgTestClient } from 'pgsql-test/test-client';
import type { SeedAdapter } from 'pgsql-test/seed/types';
import type {
  GraphQLQueryOptions,
  GraphQLQueryFnObj,
  GraphQLQueryFn,
  GraphQLQueryUnwrappedFnObj,
  GraphQLQueryUnwrappedFn,
  GraphQLResponse,
  GetConnectionsInput
} from './types';

// Core unwrapping utility
const unwrap = <T>(res: GraphQLResponse<T>): T => {
  if (res.errors?.length) {
    throw new Error(JSON.stringify(res.errors, null, 2));
  }
  if (!res.data) {
    throw new Error('No data returned from GraphQL query');
  }
  return res.data;
};

// Base connection setup - shared across all variants
const createConnectionsBase = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
) => {
  const conn: GetConnectionResult = await getPgConnections(input, seedAdapters);
  const { pg, db, teardown: dbTeardown } = conn;

  const gqlContext = GraphQLTest(input, conn);
  await gqlContext.setup();

  const teardown = async () => {
    await gqlContext.teardown();
    await dbTeardown();
  };

  const baseQuery = (opts: GraphQLQueryOptions) => gqlContext.query(opts);
  const baseQueryPositional = (query: any, variables?: any, commit?: boolean, reqOptions?: any) =>
    gqlContext.query({ query, variables, commit, reqOptions });

  return {
    pg,
    db,
    teardown,
    baseQuery,
    baseQueryPositional
  };
};

// ============================================================================
// REGULAR QUERY VERSIONS
// ============================================================================

/**
 * Creates connections with raw GraphQL responses
 */
export const getConnectionsObject = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFnObj;
}> => {
  const { pg, db, teardown, baseQuery } = await createConnectionsBase(input, seedAdapters);

  return {
    pg,
    db,
    teardown,
    query: baseQuery
  };
};

/**
 * Creates connections with unwrapped GraphQL responses (throws on errors)
 */
export const getConnectionsObjectUnwrapped = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryUnwrappedFnObj;
}> => {
  const { pg, db, teardown, baseQuery } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryUnwrappedFnObj = async (opts) => unwrap(await baseQuery(opts));

  return {
    pg,
    db,
    teardown,
    query
  };
};

/**
 * Creates connections with logging for GraphQL queries
 */
export const getConnectionsObjectWithLogging = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFnObj;
}> => {
  const { pg, db, teardown, baseQuery } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryFnObj = async (opts) => {
    console.log('Executing GraphQL query:', opts.query);
    const result = await baseQuery(opts);
    console.log('GraphQL result:', result);
    return result;
  };

  return {
    pg,
    db,
    teardown,
    query
  };
};

/**
 * Creates connections with timing for GraphQL queries
 */
export const getConnectionsObjectWithTiming = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFnObj;
}> => {
  const { pg, db, teardown, baseQuery } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryFnObj = async (opts) => {
    const start = Date.now();
    const result = await baseQuery(opts);
    const duration = Date.now() - start;
    console.log(`GraphQL query took ${duration}ms`);
    return result;
  };

  return {
    pg,
    db,
    teardown,
    query
  };
};

// ============================================================================
// POSITIONAL QUERY VERSIONS
// ============================================================================

/**
 * Creates connections with raw GraphQL responses (positional API)
 */
export const getConnections = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFn;
}> => {
  const { pg, db, teardown, baseQueryPositional } = await createConnectionsBase(input, seedAdapters);

  return {
    pg,
    db,
    teardown,
    query: baseQueryPositional
  };
};

/**
 * Creates connections with unwrapped GraphQL responses (positional API, throws on errors)
 */
export const getConnectionsUnwrapped = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryUnwrappedFn;
}> => {
  const { pg, db, teardown, baseQueryPositional } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryUnwrappedFn = async (query, variables, commit, reqOptions) =>
    unwrap(await baseQueryPositional(query, variables, commit, reqOptions));

  return {
    pg,
    db,
    teardown,
    query
  };
};

/**
 * Creates connections with logging for GraphQL queries (positional API)
 */
export const getConnectionsWithLogging = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFn;
}> => {
  const { pg, db, teardown, baseQueryPositional } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryFn = async (query, variables, commit, reqOptions) => {
    console.log('Executing positional GraphQL query:', query);
    const result = await baseQueryPositional(query, variables, commit, reqOptions);
    console.log('GraphQL result:', result);
    return result;
  };

  return {
    pg,
    db,
    teardown,
    query
  };
};

/**
 * Creates connections with timing for GraphQL queries (positional API)
 */
export const getConnectionsWithTiming = async (
  input: GetConnectionsInput & GetConnectionOpts,
  seedAdapters?: SeedAdapter[]
): Promise<{
  pg: PgTestClient;
  db: PgTestClient;
  teardown: () => Promise<void>;
  query: GraphQLQueryFn;
}> => {
  const { pg, db, teardown, baseQueryPositional } = await createConnectionsBase(input, seedAdapters);

  const query: GraphQLQueryFn = async (query, variables, commit, reqOptions) => {
    const start = Date.now();
    const result = await baseQueryPositional(query, variables, commit, reqOptions);
    const duration = Date.now() - start;
    console.log(`Positional GraphQL query took ${duration}ms`);
    return result;
  };

  return {
    pg,
    db,
    teardown,
    query
  };
};