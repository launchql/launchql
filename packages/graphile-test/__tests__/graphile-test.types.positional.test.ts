process.env.LOG_SCOPE = 'graphile-test';

import { snapshot } from '../src';
import { seed } from 'pgsql-test';
import { join } from 'path';
import type { GraphQLQueryFn } from '../src/types';
import type { PgTestClient } from 'pgsql-test/test-client';
import { getConnections } from '../src/get-connections';

const schemas = ['app_public'];
const sql = (f: string) => join(__dirname, '/../sql', f);

let teardown: () => Promise<void>;
let query: GraphQLQueryFn;
let db: PgTestClient;

beforeAll(async () => {
    const connections = await getConnections(
        {
            schemas,
            authRole: 'authenticated'
        },
        [
            seed.sqlfile([
                sql('test.sql'),
                sql('grants.sql')
            ])
        ]
    );

    ({ query, db, teardown } = connections);
});

beforeEach(() => db.beforeEach());
afterEach(() => db.afterEach());
afterAll(() => teardown());

it('creates a user and returns typed result', async () => {
    db.setContext({
        role: 'authenticated'
    })
    interface CreateUserVariables {
        input: {
            user: {
                username: string;
            };
        };
    }

    interface CreateUserResult {
        createUser: {
            user: {
                id: number;
                username: string;
            };
        };
    }

    const createUserMutation = `
        mutation CreateUser($input: CreateUserInput!) {
          createUser(input: $input) {
            user {
              id
              username
            }
          }
        }
    `;

    const variables: CreateUserVariables = {
        input: {
            user: {
                username: 'alice'
            }
        }
    };

    // Using positional API: query, variables, commit, reqOptions
    const result = await query<CreateUserResult>(createUserMutation, variables);

    // Assertions - need to check .data since this is NOT unwrapped
    expect(result).toBeDefined();
    expect(result.errors).toBeUndefined(); // Check no errors
    expect(result.data).toBeDefined();
    expect(result.data!.createUser).toBeDefined();
    expect(result.data!.createUser.user.username).toBe('alice');
    expect(typeof result.data!.createUser.user.id).toBe('number');

    // Optional snapshot for structure
    expect(snapshot(result.data)).toMatchSnapshot('create-user');
});

it('handles errors gracefully with raw GraphQL responses', async () => {
    db.setContext({
        role: 'authenticated'
    });

    interface CreateUserVariables {
        input: {
            user: {
                username: string;
            };
        };
    }

    interface CreateUserResult {
        createUser: {
            user: {
                id: number;
                username: string;
            };
        };
    }

    const createUserMutation = `
        mutation CreateUser($input: CreateUserInput!) {
          createUser(input: $input) {
            user {
              id
              username
            }
          }
        }
    `;

    const bobVariables: CreateUserVariables = {
        input: {
            user: {
                username: 'bob'
            }
        }
    };

    // First user creation should succeed - using positional API
    const firstResult = await query<CreateUserResult>(createUserMutation, bobVariables);

    expect(firstResult.errors).toBeUndefined();
    expect(firstResult.data).toBeDefined();
    expect(firstResult.data!.createUser.user.username).toBe('bob');

    // Second user creation with same username should return errors (not throw)
    // (assuming username has a unique constraint)
    const secondResult = await query<CreateUserResult>(createUserMutation, bobVariables);

    // With raw responses, we get errors in the response instead of exceptions
    expect(secondResult.errors).toBeDefined();
    expect(secondResult.errors!.length).toBeGreaterThan(0);
    
    // Data might be partial (with null fields) rather than completely null
    expect(secondResult.data).toBeDefined();
    expect(secondResult.data!.createUser).toBeNull(); // The mutation result should be null
    
    // We can inspect the actual error message
    expect(secondResult.errors![0].message).toContain('duplicate'); // or whatever your constraint error says
});