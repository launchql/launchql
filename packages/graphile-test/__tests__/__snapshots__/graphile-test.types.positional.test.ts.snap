// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates a user and returns typed result: create-user 1`] = `
{
  "createUser": {
    "user": {
      "id": "[ID]",
      "username": "alice",
    },
  },
}
`;
